src/mame/drivers/firebeat.cpp

```diff
@@ -865,368 +1127,381 @@ WRITE_LINE_MEMBER(firebeat_state::spu_ata_dmarq)
+	if (state == 0)
+		m_audiocpu->set_input_line(INPUT_LINE_IRQ2, state);
 
-	/* configure fast RAM regions for DRC */
-	m_maincpu->ppcdrc_add_fastram(0x00000000, 0x01ffffff, false, m_work_ram);
+	m_sync_ata_irq = state;
 }
 
-void firebeat_state::firebeat_map(address_map &map)
+TIMER_DEVICE_CALLBACK_MEMBER(firebeat_spu_state::spu_timer_callback)
 {
-	map(0x00000000, 0x01ffffff).ram().share("work_ram");
-	map(0x70000000, 0x70000fff).rw(FUNC(firebeat_state::midi_uart_r), FUNC(firebeat_state::midi_uart_w)).umask32(0xff000000);
-	map(0x70006000, 0x70006003).w(FUNC(firebeat_state::extend_board_irq_w));
-	map(0x70008000, 0x7000800f).r(FUNC(firebeat_state::keyboard_wheel_r));
-	map(0x7000a000, 0x7000a003).r(FUNC(firebeat_state::extend_board_irq_r));
-	map(0x74000000, 0x740003ff).noprw(); // SPU shared RAM
-	map(0x7d000200, 0x7d00021f).r(FUNC(firebeat_state::cabinet_r));
-	map(0x7d000340, 0x7d000347).r(FUNC(firebeat_state::sensor_r));
-	map(0x7d000400, 0x7d000401).rw("ymz", FUNC(ymz280b_device::read), FUNC(ymz280b_device::write));
-	map(0x7d000800, 0x7d000803).r(FUNC(firebeat_state::input_r));
-	map(0x7d400000, 0x7d5fffff).rw("flash_main", FUNC(fujitsu_29f016a_device::read), FUNC(fujitsu_29f016a_device::write));
-	map(0x7d800000, 0x7d9fffff).rw("flash_snd1", FUNC(fujitsu_29f016a_device::read), FUNC(fujitsu_29f016a_device::write));
-	map(0x7da00000, 0x7dbfffff).rw("flash_snd2", FUNC(fujitsu_29f016a_device::read), FUNC(fujitsu_29f016a_device::write));
-	map(0x7dc00000, 0x7dc0000f).rw(m_duart_com, FUNC(pc16552_device::read), FUNC(pc16552_device::write));
-	map(0x7e000000, 0x7e00003f).rw("rtc", FUNC(rtc65271_device::rtc_r), FUNC(rtc65271_device::rtc_w));
-	map(0x7e000100, 0x7e00013f).rw("rtc", FUNC(rtc65271_device::xram_r), FUNC(rtc65271_device::xram_w));
-	map(0x7e800000, 0x7e8000ff).rw(m_gcu[0], FUNC(k057714_device::read), FUNC(k057714_device::write));
-	map(0x7fe00000, 0x7fe0000f).rw(FUNC(firebeat_state::ata_command_r), FUNC(firebeat_state::ata_command_w));
-	map(0x7fe80000, 0x7fe8000f).rw(FUNC(firebeat_state::ata_control_r), FUNC(firebeat_state::ata_control_w));
-	map(0x7ff80000, 0x7fffffff).rom().region("user1", 0);       /* System BIOS */
+	if (m_sync_ata_irq)
+		m_audiocpu->set_input_line(INPUT_LINE_IRQ6, 1);
+	else
+		m_audiocpu->set_input_line(INPUT_LINE_IRQ2, 1);
 }
 
-void firebeat_state::firebeat_spu_map(address_map &map)
+/*****************************************************************************
+* beatmania III
+******************************************************************************/
+void firebeat_bm3_state::firebeat_bm3(machine_config &config)
 {
-	firebeat_map(map);
-	map(0x74000000, 0x740003ff).rw(m_dpram, FUNC(cy7c131_device::right_r), FUNC(cy7c131_device::right_w)); // SPU shared RAM
+	firebeat_spu_base(config);
+
+	m_maincpu->set_addrmap(AS_PROGRAM, &firebeat_bm3_state::firebeat_bm3_map);
+
+	screen_device *screen = subdevice<screen_device>("screen");
+	if (screen != nullptr) {
```

Comment:
Is it possible to end up here without a screen?  It always calls `firebeat_spu_base` which I’d expect to either have a screen or not have a screen (from what I can see it always has one, instantiated in `firebeat`).  It shouldn’t need the test.

Reply:
No, a screen should always exist but I guarded it out of habit since it wasn't in a tight loop. If the screen doesn't exist at that point then there's probably bigger issues that need to be dealt with I guess so the null check probably isn't all that useful I guess.